tests/xen/test_ring0.py

```diff
@@ -0,0 +1,135 @@
+import secrets
+import time
+from typing import Optional
+import pytest
+
+from lib.commands import SSHCommandFailed
+from lib.host import Host
+from lib.vm import VM
+
+# Requirements:
+# From --hosts parameter:
+# - host: XCP-ng host >= 8.2, reboots after test
+# From --vm parameter:
+# - A VM to import
+# Only XST tests with pass/fail results, and that don't crash the host were included.
+
+
+def do_execute_xst(host: Host, modname: str, testname: Optional[str] = None):
+    if testname is None:
+        testname = modname
+    host.ssh(f"modprobe xst_{modname}")
+    try:
+        host.ssh(f"echo 1 > /sys/kernel/debug/xst/{testname}/run")
+        host.ssh(f"grep -q 'status: pass' /sys/kernel/debug/xst/{testname}/results")
+    finally:
+        host.ssh(f"modprobe -r xst_{modname}", check=False)
+
+
+@pytest.mark.reboot  # host_with_ring0_tests
+@pytest.mark.usefixtures("host_with_ring0_tests")
+class TestRing0Tests:
+    def test_privcmd_restrict(self, host: Host):
+        host.ssh("/usr/bin/privcmd-restrict_test")
+
+    def test_xst_alloc_balloon(self, host: Host):
+        do_execute_xst(host, "alloc_balloon")
+
+    def test_xst_big_module(self, host: Host):
+        do_execute_xst(host, "big_module")
+
+    def test_xst_evtchn_latency(self, host: Host):
+        do_execute_xst(host, "evtchn_latency", "evtchn_lat")
+
+    @pytest.mark.skip("only makes sense for 2l evtchn")
+    def test_xst_evtchn_limit(self, host: Host):
+        do_execute_xst(host, "evtchn_limit")
+
+    def test_xst_evtchn_stress(self, host: Host):
+        do_execute_xst(host, "evtchn_stress")
+
+    @pytest.mark.skip("leaks event channels infinitely")
+    def test_xst_evtchn_unbind(self, host: Host):
+        do_execute_xst(host, "evtchn_unbind")
+
+    def test_xst_get_user_pages(self, host: Host):
+        host.ssh("modprobe xst_get_user_pages")
+        try:
+            host.ssh("/usr/bin/gup_test")
+        finally:
+            host.ssh("modprobe -r xst_get_user_pages", check=False)
+
+    def test_xst_grant_copy_perf(self, host: Host):
+        do_execute_xst(host, "grant_copy_perf", "gntcpy_perf")
+
+    @pytest.mark.small_vm
+    def test_xst_ioemu_msi(self, host: Host, running_unix_vm: VM):
+        # TODO: validate MSI reception in guest
+        vm = running_unix_vm
+        domid = vm.param_get("dom-id")
+        host.ssh("modprobe xst_ioemu_msi")
+        try:
+            host.ssh(f"echo {domid} > /sys/kernel/debug/xst/ioemu_msi/domid")
+            host.ssh("echo 1 > /sys/kernel/debug/xst/ioemu_msi/data")
+            host.ssh("echo 1 > /sys/kernel/debug/xst/ioemu_msi/run")
+            host.ssh("grep -q 'status: pass' /sys/kernel/debug/xst/ioemu_msi/results")
+        finally:
+            host.ssh("modprobe -r xst_ioemu_msi", check=False)
+
+    def test_xst_livepatch(self, host: Host):
+        try:
+            host.ssh("lsmod | grep -wq livepatch_tester")
+            pytest.skip("livepatch_tester already loaded, needs reboot")
```

Comment:
We avoid skipping tests for such reasons. One might falsely think that the test was executed, in CI, as no one checks all skips one by one. What we do instead is check for prerequisites in a fixture and fail the test if the prerequisite is not met.

Reply:
I would not use `fail`, that's rather what `error` is for

Reply:
I couldn't find `pytest.error` anywhere.

Reply:
By moving the prerequisite test to a fixture as I suggested (and that's one of the things fixtures are here for in pytest), it would be obvious that the failure is in the prerequisites and not in the test itself.

Reply:
Yes, my bad, there does not seem to be another mechanism than fixtures to generate error rather than failure (https://docs.pytest.org/en/stable/explanation/fixtures.html#fixture-errors seems to be where they talk about it, but it does not spell out precisely that errors are solely coming from fixtures)

Reply:
Next question would be, do we want to do this in a `host_without_livepatch_tester` fixture to be able to get an error rather than a failure?

Reply:
That's precisely what I'm talking about from the beginning :D (though in other fixtures we do use pytest.fail too at the moment)

Reply:
Fixed with the suggested fixture.